atlas-engineer/nyxt
# WebExtensions API support

This adds a rudimentary suport for the WebExtensions API used to
create extensions for popular browsers (Chrome, Firefox, Opera etc.)

While not immediately useful, this is a good prototype to start the
WebExtensions support in Nyxt and to discuss the design decisions made
so far (I've been working on it alone, and some bug/quirk may have
slipped through without me noticing).

# What's new

- source/web-extensions.lisp: Lisp-side support for WebExtensions --
  mainly manifest.json parsing and smart file search (it's not that
  WebExtensions require much more than that on the browsers side.)
  
- source/renderers/gtk.lisp: Numerous hacks to make different APIs
  work as standardized. WebKit is not exactly WebExtensions-ready, so
  there's quite a lot of nasty code around.
  
- libraries/web-extensions: A directory with a
  [WebKitExtension](https://webkitgtk.org/reference/webkit2gtk/stable/WebKitWebExtension.html)
  library creating numerous JS APIs and communicating with Lisp side
  via [WebKitUserMessages](https://webkitgtk.org/reference/webkit2gtk/stable/WebKitUserMessage.html)

- Small typo/style/API changes wherever I was
  lazy/overwhelmed/hesitating to push those onto master.

# Known Issues

## Security

This is terrible implementation from the viewpoint of security. Issues are (most are fixed, the remaining ones are either not fixable or minor):

- [X] Extensions' permissions are not restricted at all, the corresponding
  manifest.json key is not even looked at.
  - Done.

- [ ] runtime.getURL returns data: URLs which are only suitable for
  resource loading, while [being a security hole when navigated
  to](https://blog.mozilla.org/security/2017/11/27/blocking-top-level-navigations-data-urls-firefox-59/).

- [x] Custom scheme handling is extremely fragile in general.
  - #1815.

- [X] The privileged extension pages (i.e. background, popup, options
  pages) assign the privileges to the first privileged extension they
  are passed. This may not be the extension they actually belong
  to.
  - I actually fixed it -- now Lisp side passes only one extension, while C side only does privileges for the first privileged extensions it sees.

- [X] Panel buffers inherit from web buffers now, and this is not really safe. It's also breaking everything else in panel-buffers.
  - Should be fixed with #1815 and #1805.

- [ ] Memory is not freed everywhere it needs to be. This can clutter our memory with time. Thus, C-side of things asks for more memory discipline.

## Incompleteness

UPDATE (Nov 6th) uBlock Origin is to be used as a benchmark for completeness of the API support, thus all the JS APIs checkboxes are replaced with uBlock-required APIs. For the full list of implemented APIs, see the README of the web-extensions library.

This is by no means the complete implementation. Aside from security,
there are things that are simply not yet done. While most of these are
optional or even browser-specific, some are must-haves. From the most
necessary to the least necessary:
- [ ] Commands.
- [ ] Option pages.
- [ ] Supporting more manifest.json keys (alphabetic order):
  - [X] browser_action
  - [X] content_scripts
  - [X] homepage_url
  - [X] icons
  - [X] manifest_version
  - [ ] author
  - [ ] background
  - [ ] browser_specific_settings
  - [ ] chrome_settings_overrides
  - [ ] chrome_url_overrides
  - [ ] commands
  - [ ] content_security_policy
  - [ ] default_locale
  - [ ] developer
  - [ ] devtools_page
  - [ ] dictionaries
  - [ ] externally_connectable
  - [ ] incognito
  - [ ] offline_enabled
  - [ ] omnibox
  - [ ] optional_permissions
  - [ ] options_page
  - [ ] options_ui
  - [ ] page_action
  - [ ] protocol_handlers
  - [ ] short_name
  - [ ] sidebar_action
  - [ ] storage
  - [ ] theme
  - [ ] theme_experiment
  - [ ] user_scripts
  - [ ] version_name
  - [ ] web_accessible_resources
- Supporting uBlock-used APIs:
  - Methods (alphabetically, with unimplemented on top):
    - [ ] browser.browserAction.setPopup -- easy
    - [ ] browser.commands.getAll -- easy
    - [ ] browser.commands.reset -- easy
    - [ ] browser.commands.update -- easy
    - [ ] browser.i18n.getMessage -- easy?
    - [ ] browser.runtime.connect
    - [ ] browser.runtime.reload
    - [ ] browser.tabs.move -- easy
    - [ ] browser.webRequest.filterResponseData
    - [ ] browser.webRequest.handlerBehaviorChanged
    - [X] browser.runtime.getBrowserInfo
    - [X] browser.runtime.getManifest
    - [X] browser.runtime.getURL
    - [X] browser.tabs.create
  - Variables and types (will be mostly easy to do):
    - [ ] browser.runtime.lastError
    - [ ] browser.windows.WINDOW_ID_NONE -- easy
    - [ ] browser.webRequest.ResourceType -- easy
  - Events (alphabetically, with unimplemented on top, hardest to do as they require modifications all over Nyxt sources):
    - [ ] browser.browserAction.onClicked -- easy
    - [ ] browser.commands.onCommand -- easy
    - [ ] browser.runtime.onConnect
    - [ ] browser.runtime.onUpdateAvailable
    - [ ] browser.tabs.onActivated
    - [ ] browser.tabs.onCreated -- easy
    - [ ] browser.tabs.onRemoved -- easy
    - [ ] browser.tabs.onUpdated -- easy
    - [ ] browser.webNavigation.onCommitted
    - [ ] browser.webNavigation.onCreatedNavigationTarget
    - [ ] browser.webRequest.onBeforeRequest
    - [ ] browser.webRequest.onHeadersReceived
    - [ ] browser.windows.onFocusChanged -- easy
  - Non-mandatory but cool APIs because uBlock uses them to provide more features:
    - [ ] browser.dns
    - [ ] browser.privacy
    - [ ] browser.menus 
      
## WebKitGTK incompleteness

@jmercouris recently said he thought we'd need to patch WebKit to
enable WebExtensions support. While this is not (yet) the case, I've
been dangerously close to digging into how to contribute to WebKit.

The reason is that Apple's original WebKit has [a way to load resource from custom schemes](https://developer.apple.com/documentation/webkit/wkurlschemehandler), while WebKitGTK [does not support that](https://bugs.webkit.org/show_bug.cgi?id=61007) and has no bindings for the WKURLSchemeHandler (link above).

# How To Test This
- Compile the C library having WebKit and all of its dependencies in the path. 
  - In case of Guix, `guix environment webkitgtk --ad-hoc webkitgtk -- make all` while in `libraries/web-extensions/` should cut it.
- Have cl-webkit updated to the latest master version!
- Check out the https://github.com/mdn/webextensions-examples
- Put this code into yout init.lisp:
``` lisp
(define-configuration web-buffer
  ((gtk-extensions-path
    (make-instance 'gtk-extensions-data-path
                   :dirname (asdf:system-relative-pathname :nyxt "libraries/web-extensions")))))

(nyxt/web-extensions:load-web-extension beastify "/absolute/path/to/webextensions-examples/beastify/")

(define-configuration web-buffer
  ((default-modes `(beastify ,@%slot-default%))))
```
- Run Nyxt
- Navigate to https://www.mozilla.org
- Click the turtle icon in the mode line or "Beastify" mode name.
- Try clicking buttons in the panel buffer opened.
- The extension should work fine -- animal buttons insert animals, reset button removing the beasts.

Closes #1758.
Closes #1670.

EDIT: Add checkboxes to security points.
EDIT: Check the completed security points.
EDIT: Checkboxes specifically for uBlock.
